From 187844389881597ab4ca46ea56ecdefcdc5e12a9 Mon Sep 17 00:00:00 2001
From: Guilhem Bichot <guilhem.bichot@contributor>

Date: Mon, 4 Aug 2025 19:00:40 +0200
Subject: [PATCH] while using windown function and order by
 together, after enable plan_cache, the query using cached plan will return a
 result set of incorrectly sorted

when caching the plan, we missed copying JOIN::m_windows_sort . As a
result, the query's top ORDER BY got lost on second execution in
JOIN::create_intermediate_table() (see 'order.clean()' there: the
ORDER-BY was pushed to before-windows, but as windows actually sorted
rows, it means this ORDER-BY was actually lost). So rows were emitted
in wrong order.
---
 mysql-test/include/session_plan_cache.inc     |  55 ++
 mysql-test/r/session_plan_cache.result        | 490 ++++++++++++++++++
 .../dstore_main/r/session_plan_cache.result   | 490 ++++++++++++++++++
 sql/sql_plan_cache.cc                         |   1 +
 4 files changed, 1036 insertions(+)

diff --git a/mysql-test/include/session_plan_cache.inc b/mysql-test/include/session_plan_cache.inc
index 00d07eb8d09..a92f7de4104 100644
--- a/mysql-test/include/session_plan_cache.inc
+++ b/mysql-test/include/session_plan_cache.inc
@@ -1593,4 +1593,59 @@ execute s using @a;
 drop prepare s;
 drop table t2;
 
+--echo #
+--echo # BUGFIX while using windown function and order by together, after enable plan_cache, the query using cached plan will return a result set of incorrectly sorted
+--echo #
+
+CREATE TABLE t_general_test_pk (
+  id int NOT NULL auto_increment primary key,
+  name varchar(255) DEFAULT NULL,
+  client char(4) DEFAULT NULL,
+  bin_id binary(16) NOT NULL,
+  t_bit16 bit(16) DEFAULT NULL,
+  create_time datetime DEFAULT CURRENT_TIMESTAMP COMMENT '????',
+  col1 decimal(10,2) DEFAULT NULL,
+  text blob,
+  tex text,
+  sex enum('boy','girl','secret') DEFAULT 'secret'
+) DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
+
+insert into t_general_test_pk values(1,null,null,123,0,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(2,null,null,123,0,'2025-08-04 18:42:00',0,null,null,'boy');
+insert into t_general_test_pk values(3,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(4,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'','','girl');
+insert into t_general_test_pk values(5,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(6,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'','','girl');
+insert into t_general_test_pk values(7,'NULL','NULL',123,2,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(8,'NULL','NULL',123,2,'2025-08-04 18:42:00',0,'','','girl');
+insert into t_general_test_pk values(9,'NULL','NULL',123,2,'2025-08-04 18:42:00',0,'','','secret');
+
+insert into t_general_test_pk(name,client,bin_id,t_bit16,create_time,col1,text,tex,sex) select name,client,bin_id,t_bit16,create_time,col1,text,tex,sex from t_general_test_pk;
+insert into t_general_test_pk(name,client,bin_id,t_bit16,create_time,col1,text,tex,sex) select name,client,bin_id,t_bit16,create_time,col1,text,tex,sex from t_general_test_pk;
+insert into t_general_test_pk(name,client,bin_id,t_bit16,create_time,col1,text,tex,sex) select name,client,bin_id,t_bit16,create_time,col1,text,tex,sex from t_general_test_pk;
+
+create index id_key on t_general_test_pk(id desc);
+create index name_key on t_general_test_pk(name(20) desc, create_time);
+create index decimal_key on t_general_test_pk(col1);
+create index sex_key on t_general_test_pk(sex desc);
+
+# used to lose ORDER BY
+
+set @sql_text1='select sex,create_time,id,id- LAG(id) over (order by create_time) as temp_diff_id from t_general_test_pk t order by sex,create_time;';
+PREPARE stmt1 FROM @sql_text1;
+EXECUTE stmt1;
+EXECUTE stmt1;
+EXECUTE stmt1;
+show status like '%Cached_plan%';
+DROP PREPARE stmt1;
+ 
+set @sql_text1='select sex,create_time,id,avg(id) over (order by create_time ROWS BETWEEN 1 PRECEDING AND CURRENT ROW) as moving_avg_id from t_general_test_pk t order by sex,create_time;';
+PREPARE stmt1 FROM @sql_text1;
+EXECUTE stmt1;
+EXECUTE stmt1;
+EXECUTE stmt1;
+show status like '%Cached_plan%';
+DROP PREPARE stmt1;
+DROP TABLE t_general_test_pk;
+
 set rds_plan_cache = @old_plan_cache;
diff --git a/mysql-test/r/session_plan_cache.result b/mysql-test/r/session_plan_cache.result
index 5a8e77e73cd..0859b5faa6b 100644
--- a/mysql-test/r/session_plan_cache.result
+++ b/mysql-test/r/session_plan_cache.result
@@ -2818,4 +2818,494 @@ Warnings:
 Note	1003	/* select#1 */ select `test`.`t2`.`pk` AS `pk`,`test`.`t2`.`a` AS `a`,`test`.`t2`.`b` AS `b` from `test`.`t2` where (`test`.`t2`.`a` = 3)
 drop prepare s;
 drop table t2;
+#
+# BUGFIX while using windown function and order by together, after enable plan_cache, the query using cached plan will return a result set of incorrectly sorted
+#
+CREATE TABLE t_general_test_pk (
+id int NOT NULL auto_increment primary key,
+name varchar(255) DEFAULT NULL,
+client char(4) DEFAULT NULL,
+bin_id binary(16) NOT NULL,
+t_bit16 bit(16) DEFAULT NULL,
+create_time datetime DEFAULT CURRENT_TIMESTAMP COMMENT '????',
+col1 decimal(10,2) DEFAULT NULL,
+text blob,
+tex text,
+sex enum('boy','girl','secret') DEFAULT 'secret'
+) DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
+insert into t_general_test_pk values(1,null,null,123,0,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(2,null,null,123,0,'2025-08-04 18:42:00',0,null,null,'boy');
+insert into t_general_test_pk values(3,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(4,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'','','girl');
+insert into t_general_test_pk values(5,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(6,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'','','girl');
+insert into t_general_test_pk values(7,'NULL','NULL',123,2,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(8,'NULL','NULL',123,2,'2025-08-04 18:42:00',0,'','','girl');
+insert into t_general_test_pk values(9,'NULL','NULL',123,2,'2025-08-04 18:42:00',0,'','','secret');
+insert into t_general_test_pk(name,client,bin_id,t_bit16,create_time,col1,text,tex,sex) select name,client,bin_id,t_bit16,create_time,col1,text,tex,sex from t_general_test_pk;
+insert into t_general_test_pk(name,client,bin_id,t_bit16,create_time,col1,text,tex,sex) select name,client,bin_id,t_bit16,create_time,col1,text,tex,sex from t_general_test_pk;
+insert into t_general_test_pk(name,client,bin_id,t_bit16,create_time,col1,text,tex,sex) select name,client,bin_id,t_bit16,create_time,col1,text,tex,sex from t_general_test_pk;
+create index id_key on t_general_test_pk(id desc);
+create index name_key on t_general_test_pk(name(20) desc, create_time);
+create index decimal_key on t_general_test_pk(col1);
+create index sex_key on t_general_test_pk(sex desc);
+set @sql_text1='select sex,create_time,id,id- LAG(id) over (order by create_time) as temp_diff_id from t_general_test_pk t order by sex,create_time;';
+PREPARE stmt1 FROM @sql_text1;
+EXECUTE stmt1;
+sex	create_time	id	temp_diff_id
+boy	2025-08-04 18:42:00	5	-1
+boy	2025-08-04 18:42:00	56	-28
+boy	2025-08-04 18:42:00	80	-1
+boy	2025-08-04 18:42:00	78	-1
+boy	2025-08-04 18:42:00	76	-1
+boy	2025-08-04 18:42:00	75	-1
+boy	2025-08-04 18:42:00	65	NULL
+boy	2025-08-04 18:42:00	16	-1
+boy	2025-08-04 18:42:00	14	-1
+boy	2025-08-04 18:42:00	12	-1
+boy	2025-08-04 18:42:00	11	-1
+boy	2025-08-04 18:42:00	25	14
+boy	2025-08-04 18:42:00	7	-1
+boy	2025-08-04 18:42:00	10	-65
+boy	2025-08-04 18:42:00	3	-1
+boy	2025-08-04 18:42:00	2	-1
+boy	2025-08-04 18:42:00	34	32
+boy	2025-08-04 18:42:00	40	-1
+boy	2025-08-04 18:42:00	38	-1
+boy	2025-08-04 18:42:00	36	-1
+boy	2025-08-04 18:42:00	35	-1
+boy	2025-08-04 18:42:00	1	-34
+boy	2025-08-04 18:42:00	31	-1
+boy	2025-08-04 18:42:00	29	-1
+boy	2025-08-04 18:42:00	27	-1
+boy	2025-08-04 18:42:00	26	-1
+boy	2025-08-04 18:42:00	67	-1
+boy	2025-08-04 18:42:00	62	-1
+boy	2025-08-04 18:42:00	84	-1
+boy	2025-08-04 18:42:00	60	-1
+boy	2025-08-04 18:42:00	74	8
+boy	2025-08-04 18:42:00	58	-1
+boy	2025-08-04 18:42:00	57	-1
+boy	2025-08-04 18:42:00	83	26
+boy	2025-08-04 18:42:00	66	-1
+boy	2025-08-04 18:42:00	69	-1
+boy	2025-08-04 18:42:00	89	-1
+boy	2025-08-04 18:42:00	87	-1
+boy	2025-08-04 18:42:00	71	-1
+boy	2025-08-04 18:42:00	85	-1
+girl	2025-08-04 18:42:00	68	-1
+girl	2025-08-04 18:42:00	6	-1
+girl	2025-08-04 18:42:00	37	-1
+girl	2025-08-04 18:42:00	32	-1
+girl	2025-08-04 18:42:00	70	-1
+girl	2025-08-04 18:42:00	39	-1
+girl	2025-08-04 18:42:00	30	-1
+girl	2025-08-04 18:42:00	41	-1
+girl	2025-08-04 18:42:00	63	-1
+girl	2025-08-04 18:42:00	28	-1
+girl	2025-08-04 18:42:00	72	-1
+girl	2025-08-04 18:42:00	61	-1
+girl	2025-08-04 18:42:00	81	-1
+girl	2025-08-04 18:42:00	86	-1
+girl	2025-08-04 18:42:00	79	-1
+girl	2025-08-04 18:42:00	77	-1
+girl	2025-08-04 18:42:00	88	-1
+girl	2025-08-04 18:42:00	90	-1
+girl	2025-08-04 18:42:00	17	-1
+girl	2025-08-04 18:42:00	15	-1
+girl	2025-08-04 18:42:00	13	-1
+girl	2025-08-04 18:42:00	59	-1
+girl	2025-08-04 18:42:00	8	-1
+girl	2025-08-04 18:42:00	4	-1
+secret	2025-08-04 18:42:00	91	8
+secret	2025-08-04 18:42:00	18	8
+secret	2025-08-04 18:42:00	33	32
+secret	2025-08-04 18:42:00	9	-16
+secret	2025-08-04 18:42:00	42	8
+secret	2025-08-04 18:42:00	64	-10
+secret	2025-08-04 18:42:00	82	26
+secret	2025-08-04 18:42:00	73	8
+EXECUTE stmt1;
+sex	create_time	id	temp_diff_id
+boy	2025-08-04 18:42:00	5	-1
+boy	2025-08-04 18:42:00	56	-28
+boy	2025-08-04 18:42:00	80	-1
+boy	2025-08-04 18:42:00	78	-1
+boy	2025-08-04 18:42:00	76	-1
+boy	2025-08-04 18:42:00	75	-1
+boy	2025-08-04 18:42:00	65	NULL
+boy	2025-08-04 18:42:00	16	-1
+boy	2025-08-04 18:42:00	14	-1
+boy	2025-08-04 18:42:00	12	-1
+boy	2025-08-04 18:42:00	11	-1
+boy	2025-08-04 18:42:00	25	14
+boy	2025-08-04 18:42:00	7	-1
+boy	2025-08-04 18:42:00	10	-65
+boy	2025-08-04 18:42:00	3	-1
+boy	2025-08-04 18:42:00	2	-1
+boy	2025-08-04 18:42:00	34	32
+boy	2025-08-04 18:42:00	40	-1
+boy	2025-08-04 18:42:00	38	-1
+boy	2025-08-04 18:42:00	36	-1
+boy	2025-08-04 18:42:00	35	-1
+boy	2025-08-04 18:42:00	1	-34
+boy	2025-08-04 18:42:00	31	-1
+boy	2025-08-04 18:42:00	29	-1
+boy	2025-08-04 18:42:00	27	-1
+boy	2025-08-04 18:42:00	26	-1
+boy	2025-08-04 18:42:00	67	-1
+boy	2025-08-04 18:42:00	62	-1
+boy	2025-08-04 18:42:00	84	-1
+boy	2025-08-04 18:42:00	60	-1
+boy	2025-08-04 18:42:00	74	8
+boy	2025-08-04 18:42:00	58	-1
+boy	2025-08-04 18:42:00	57	-1
+boy	2025-08-04 18:42:00	83	26
+boy	2025-08-04 18:42:00	66	-1
+boy	2025-08-04 18:42:00	69	-1
+boy	2025-08-04 18:42:00	89	-1
+boy	2025-08-04 18:42:00	87	-1
+boy	2025-08-04 18:42:00	71	-1
+boy	2025-08-04 18:42:00	85	-1
+girl	2025-08-04 18:42:00	68	-1
+girl	2025-08-04 18:42:00	6	-1
+girl	2025-08-04 18:42:00	37	-1
+girl	2025-08-04 18:42:00	32	-1
+girl	2025-08-04 18:42:00	70	-1
+girl	2025-08-04 18:42:00	39	-1
+girl	2025-08-04 18:42:00	30	-1
+girl	2025-08-04 18:42:00	41	-1
+girl	2025-08-04 18:42:00	63	-1
+girl	2025-08-04 18:42:00	28	-1
+girl	2025-08-04 18:42:00	72	-1
+girl	2025-08-04 18:42:00	61	-1
+girl	2025-08-04 18:42:00	81	-1
+girl	2025-08-04 18:42:00	86	-1
+girl	2025-08-04 18:42:00	79	-1
+girl	2025-08-04 18:42:00	77	-1
+girl	2025-08-04 18:42:00	88	-1
+girl	2025-08-04 18:42:00	90	-1
+girl	2025-08-04 18:42:00	17	-1
+girl	2025-08-04 18:42:00	15	-1
+girl	2025-08-04 18:42:00	13	-1
+girl	2025-08-04 18:42:00	59	-1
+girl	2025-08-04 18:42:00	8	-1
+girl	2025-08-04 18:42:00	4	-1
+secret	2025-08-04 18:42:00	91	8
+secret	2025-08-04 18:42:00	18	8
+secret	2025-08-04 18:42:00	33	32
+secret	2025-08-04 18:42:00	9	-16
+secret	2025-08-04 18:42:00	42	8
+secret	2025-08-04 18:42:00	64	-10
+secret	2025-08-04 18:42:00	82	26
+secret	2025-08-04 18:42:00	73	8
+EXECUTE stmt1;
+sex	create_time	id	temp_diff_id
+boy	2025-08-04 18:42:00	5	-1
+boy	2025-08-04 18:42:00	56	-28
+boy	2025-08-04 18:42:00	80	-1
+boy	2025-08-04 18:42:00	78	-1
+boy	2025-08-04 18:42:00	76	-1
+boy	2025-08-04 18:42:00	75	-1
+boy	2025-08-04 18:42:00	65	NULL
+boy	2025-08-04 18:42:00	16	-1
+boy	2025-08-04 18:42:00	14	-1
+boy	2025-08-04 18:42:00	12	-1
+boy	2025-08-04 18:42:00	11	-1
+boy	2025-08-04 18:42:00	25	14
+boy	2025-08-04 18:42:00	7	-1
+boy	2025-08-04 18:42:00	10	-65
+boy	2025-08-04 18:42:00	3	-1
+boy	2025-08-04 18:42:00	2	-1
+boy	2025-08-04 18:42:00	34	32
+boy	2025-08-04 18:42:00	40	-1
+boy	2025-08-04 18:42:00	38	-1
+boy	2025-08-04 18:42:00	36	-1
+boy	2025-08-04 18:42:00	35	-1
+boy	2025-08-04 18:42:00	1	-34
+boy	2025-08-04 18:42:00	31	-1
+boy	2025-08-04 18:42:00	29	-1
+boy	2025-08-04 18:42:00	27	-1
+boy	2025-08-04 18:42:00	26	-1
+boy	2025-08-04 18:42:00	67	-1
+boy	2025-08-04 18:42:00	62	-1
+boy	2025-08-04 18:42:00	84	-1
+boy	2025-08-04 18:42:00	60	-1
+boy	2025-08-04 18:42:00	74	8
+boy	2025-08-04 18:42:00	58	-1
+boy	2025-08-04 18:42:00	57	-1
+boy	2025-08-04 18:42:00	83	26
+boy	2025-08-04 18:42:00	66	-1
+boy	2025-08-04 18:42:00	69	-1
+boy	2025-08-04 18:42:00	89	-1
+boy	2025-08-04 18:42:00	87	-1
+boy	2025-08-04 18:42:00	71	-1
+boy	2025-08-04 18:42:00	85	-1
+girl	2025-08-04 18:42:00	68	-1
+girl	2025-08-04 18:42:00	6	-1
+girl	2025-08-04 18:42:00	37	-1
+girl	2025-08-04 18:42:00	32	-1
+girl	2025-08-04 18:42:00	70	-1
+girl	2025-08-04 18:42:00	39	-1
+girl	2025-08-04 18:42:00	30	-1
+girl	2025-08-04 18:42:00	41	-1
+girl	2025-08-04 18:42:00	63	-1
+girl	2025-08-04 18:42:00	28	-1
+girl	2025-08-04 18:42:00	72	-1
+girl	2025-08-04 18:42:00	61	-1
+girl	2025-08-04 18:42:00	81	-1
+girl	2025-08-04 18:42:00	86	-1
+girl	2025-08-04 18:42:00	79	-1
+girl	2025-08-04 18:42:00	77	-1
+girl	2025-08-04 18:42:00	88	-1
+girl	2025-08-04 18:42:00	90	-1
+girl	2025-08-04 18:42:00	17	-1
+girl	2025-08-04 18:42:00	15	-1
+girl	2025-08-04 18:42:00	13	-1
+girl	2025-08-04 18:42:00	59	-1
+girl	2025-08-04 18:42:00	8	-1
+girl	2025-08-04 18:42:00	4	-1
+secret	2025-08-04 18:42:00	91	8
+secret	2025-08-04 18:42:00	18	8
+secret	2025-08-04 18:42:00	33	32
+secret	2025-08-04 18:42:00	9	-16
+secret	2025-08-04 18:42:00	42	8
+secret	2025-08-04 18:42:00	64	-10
+secret	2025-08-04 18:42:00	82	26
+secret	2025-08-04 18:42:00	73	8
+show status like '%Cached_plan%';
+Variable_name	Value
+Cached_plan_count	1
+Cached_plan_hits	10
+DROP PREPARE stmt1;
+set @sql_text1='select sex,create_time,id,avg(id) over (order by create_time ROWS BETWEEN 1 PRECEDING AND CURRENT ROW) as moving_avg_id from t_general_test_pk t order by sex,create_time;';
+PREPARE stmt1 FROM @sql_text1;
+EXECUTE stmt1;
+sex	create_time	id	moving_avg_id
+boy	2025-08-04 18:42:00	5	5.5000
+boy	2025-08-04 18:42:00	56	70.0000
+boy	2025-08-04 18:42:00	80	80.5000
+boy	2025-08-04 18:42:00	78	78.5000
+boy	2025-08-04 18:42:00	76	76.5000
+boy	2025-08-04 18:42:00	75	75.5000
+boy	2025-08-04 18:42:00	65	65.0000
+boy	2025-08-04 18:42:00	16	16.5000
+boy	2025-08-04 18:42:00	14	14.5000
+boy	2025-08-04 18:42:00	12	12.5000
+boy	2025-08-04 18:42:00	11	11.5000
+boy	2025-08-04 18:42:00	25	18.0000
+boy	2025-08-04 18:42:00	7	7.5000
+boy	2025-08-04 18:42:00	10	42.5000
+boy	2025-08-04 18:42:00	3	3.5000
+boy	2025-08-04 18:42:00	2	2.5000
+boy	2025-08-04 18:42:00	34	18.0000
+boy	2025-08-04 18:42:00	40	40.5000
+boy	2025-08-04 18:42:00	38	38.5000
+boy	2025-08-04 18:42:00	36	36.5000
+boy	2025-08-04 18:42:00	35	35.5000
+boy	2025-08-04 18:42:00	1	18.0000
+boy	2025-08-04 18:42:00	31	31.5000
+boy	2025-08-04 18:42:00	29	29.5000
+boy	2025-08-04 18:42:00	27	27.5000
+boy	2025-08-04 18:42:00	26	26.5000
+boy	2025-08-04 18:42:00	67	67.5000
+boy	2025-08-04 18:42:00	62	62.5000
+boy	2025-08-04 18:42:00	84	84.5000
+boy	2025-08-04 18:42:00	60	60.5000
+boy	2025-08-04 18:42:00	74	70.0000
+boy	2025-08-04 18:42:00	58	58.5000
+boy	2025-08-04 18:42:00	57	57.5000
+boy	2025-08-04 18:42:00	83	70.0000
+boy	2025-08-04 18:42:00	66	66.5000
+boy	2025-08-04 18:42:00	69	69.5000
+boy	2025-08-04 18:42:00	89	89.5000
+boy	2025-08-04 18:42:00	87	87.5000
+boy	2025-08-04 18:42:00	71	71.5000
+boy	2025-08-04 18:42:00	85	85.5000
+girl	2025-08-04 18:42:00	68	68.5000
+girl	2025-08-04 18:42:00	6	6.5000
+girl	2025-08-04 18:42:00	37	37.5000
+girl	2025-08-04 18:42:00	32	32.5000
+girl	2025-08-04 18:42:00	70	70.5000
+girl	2025-08-04 18:42:00	39	39.5000
+girl	2025-08-04 18:42:00	30	30.5000
+girl	2025-08-04 18:42:00	41	41.5000
+girl	2025-08-04 18:42:00	63	63.5000
+girl	2025-08-04 18:42:00	28	28.5000
+girl	2025-08-04 18:42:00	72	72.5000
+girl	2025-08-04 18:42:00	61	61.5000
+girl	2025-08-04 18:42:00	81	81.5000
+girl	2025-08-04 18:42:00	86	86.5000
+girl	2025-08-04 18:42:00	79	79.5000
+girl	2025-08-04 18:42:00	77	77.5000
+girl	2025-08-04 18:42:00	88	88.5000
+girl	2025-08-04 18:42:00	90	90.5000
+girl	2025-08-04 18:42:00	17	17.5000
+girl	2025-08-04 18:42:00	15	15.5000
+girl	2025-08-04 18:42:00	13	13.5000
+girl	2025-08-04 18:42:00	59	59.5000
+girl	2025-08-04 18:42:00	8	8.5000
+girl	2025-08-04 18:42:00	4	4.5000
+secret	2025-08-04 18:42:00	91	87.0000
+secret	2025-08-04 18:42:00	18	14.0000
+secret	2025-08-04 18:42:00	33	17.0000
+secret	2025-08-04 18:42:00	9	17.0000
+secret	2025-08-04 18:42:00	42	38.0000
+secret	2025-08-04 18:42:00	64	69.0000
+secret	2025-08-04 18:42:00	82	69.0000
+secret	2025-08-04 18:42:00	73	69.0000
+EXECUTE stmt1;
+sex	create_time	id	moving_avg_id
+boy	2025-08-04 18:42:00	5	5.5000
+boy	2025-08-04 18:42:00	56	70.0000
+boy	2025-08-04 18:42:00	80	80.5000
+boy	2025-08-04 18:42:00	78	78.5000
+boy	2025-08-04 18:42:00	76	76.5000
+boy	2025-08-04 18:42:00	75	75.5000
+boy	2025-08-04 18:42:00	65	65.0000
+boy	2025-08-04 18:42:00	16	16.5000
+boy	2025-08-04 18:42:00	14	14.5000
+boy	2025-08-04 18:42:00	12	12.5000
+boy	2025-08-04 18:42:00	11	11.5000
+boy	2025-08-04 18:42:00	25	18.0000
+boy	2025-08-04 18:42:00	7	7.5000
+boy	2025-08-04 18:42:00	10	42.5000
+boy	2025-08-04 18:42:00	3	3.5000
+boy	2025-08-04 18:42:00	2	2.5000
+boy	2025-08-04 18:42:00	34	18.0000
+boy	2025-08-04 18:42:00	40	40.5000
+boy	2025-08-04 18:42:00	38	38.5000
+boy	2025-08-04 18:42:00	36	36.5000
+boy	2025-08-04 18:42:00	35	35.5000
+boy	2025-08-04 18:42:00	1	18.0000
+boy	2025-08-04 18:42:00	31	31.5000
+boy	2025-08-04 18:42:00	29	29.5000
+boy	2025-08-04 18:42:00	27	27.5000
+boy	2025-08-04 18:42:00	26	26.5000
+boy	2025-08-04 18:42:00	67	67.5000
+boy	2025-08-04 18:42:00	62	62.5000
+boy	2025-08-04 18:42:00	84	84.5000
+boy	2025-08-04 18:42:00	60	60.5000
+boy	2025-08-04 18:42:00	74	70.0000
+boy	2025-08-04 18:42:00	58	58.5000
+boy	2025-08-04 18:42:00	57	57.5000
+boy	2025-08-04 18:42:00	83	70.0000
+boy	2025-08-04 18:42:00	66	66.5000
+boy	2025-08-04 18:42:00	69	69.5000
+boy	2025-08-04 18:42:00	89	89.5000
+boy	2025-08-04 18:42:00	87	87.5000
+boy	2025-08-04 18:42:00	71	71.5000
+boy	2025-08-04 18:42:00	85	85.5000
+girl	2025-08-04 18:42:00	68	68.5000
+girl	2025-08-04 18:42:00	6	6.5000
+girl	2025-08-04 18:42:00	37	37.5000
+girl	2025-08-04 18:42:00	32	32.5000
+girl	2025-08-04 18:42:00	70	70.5000
+girl	2025-08-04 18:42:00	39	39.5000
+girl	2025-08-04 18:42:00	30	30.5000
+girl	2025-08-04 18:42:00	41	41.5000
+girl	2025-08-04 18:42:00	63	63.5000
+girl	2025-08-04 18:42:00	28	28.5000
+girl	2025-08-04 18:42:00	72	72.5000
+girl	2025-08-04 18:42:00	61	61.5000
+girl	2025-08-04 18:42:00	81	81.5000
+girl	2025-08-04 18:42:00	86	86.5000
+girl	2025-08-04 18:42:00	79	79.5000
+girl	2025-08-04 18:42:00	77	77.5000
+girl	2025-08-04 18:42:00	88	88.5000
+girl	2025-08-04 18:42:00	90	90.5000
+girl	2025-08-04 18:42:00	17	17.5000
+girl	2025-08-04 18:42:00	15	15.5000
+girl	2025-08-04 18:42:00	13	13.5000
+girl	2025-08-04 18:42:00	59	59.5000
+girl	2025-08-04 18:42:00	8	8.5000
+girl	2025-08-04 18:42:00	4	4.5000
+secret	2025-08-04 18:42:00	91	87.0000
+secret	2025-08-04 18:42:00	18	14.0000
+secret	2025-08-04 18:42:00	33	17.0000
+secret	2025-08-04 18:42:00	9	17.0000
+secret	2025-08-04 18:42:00	42	38.0000
+secret	2025-08-04 18:42:00	64	69.0000
+secret	2025-08-04 18:42:00	82	69.0000
+secret	2025-08-04 18:42:00	73	69.0000
+EXECUTE stmt1;
+sex	create_time	id	moving_avg_id
+boy	2025-08-04 18:42:00	5	5.5000
+boy	2025-08-04 18:42:00	56	70.0000
+boy	2025-08-04 18:42:00	80	80.5000
+boy	2025-08-04 18:42:00	78	78.5000
+boy	2025-08-04 18:42:00	76	76.5000
+boy	2025-08-04 18:42:00	75	75.5000
+boy	2025-08-04 18:42:00	65	65.0000
+boy	2025-08-04 18:42:00	16	16.5000
+boy	2025-08-04 18:42:00	14	14.5000
+boy	2025-08-04 18:42:00	12	12.5000
+boy	2025-08-04 18:42:00	11	11.5000
+boy	2025-08-04 18:42:00	25	18.0000
+boy	2025-08-04 18:42:00	7	7.5000
+boy	2025-08-04 18:42:00	10	42.5000
+boy	2025-08-04 18:42:00	3	3.5000
+boy	2025-08-04 18:42:00	2	2.5000
+boy	2025-08-04 18:42:00	34	18.0000
+boy	2025-08-04 18:42:00	40	40.5000
+boy	2025-08-04 18:42:00	38	38.5000
+boy	2025-08-04 18:42:00	36	36.5000
+boy	2025-08-04 18:42:00	35	35.5000
+boy	2025-08-04 18:42:00	1	18.0000
+boy	2025-08-04 18:42:00	31	31.5000
+boy	2025-08-04 18:42:00	29	29.5000
+boy	2025-08-04 18:42:00	27	27.5000
+boy	2025-08-04 18:42:00	26	26.5000
+boy	2025-08-04 18:42:00	67	67.5000
+boy	2025-08-04 18:42:00	62	62.5000
+boy	2025-08-04 18:42:00	84	84.5000
+boy	2025-08-04 18:42:00	60	60.5000
+boy	2025-08-04 18:42:00	74	70.0000
+boy	2025-08-04 18:42:00	58	58.5000
+boy	2025-08-04 18:42:00	57	57.5000
+boy	2025-08-04 18:42:00	83	70.0000
+boy	2025-08-04 18:42:00	66	66.5000
+boy	2025-08-04 18:42:00	69	69.5000
+boy	2025-08-04 18:42:00	89	89.5000
+boy	2025-08-04 18:42:00	87	87.5000
+boy	2025-08-04 18:42:00	71	71.5000
+boy	2025-08-04 18:42:00	85	85.5000
+girl	2025-08-04 18:42:00	68	68.5000
+girl	2025-08-04 18:42:00	6	6.5000
+girl	2025-08-04 18:42:00	37	37.5000
+girl	2025-08-04 18:42:00	32	32.5000
+girl	2025-08-04 18:42:00	70	70.5000
+girl	2025-08-04 18:42:00	39	39.5000
+girl	2025-08-04 18:42:00	30	30.5000
+girl	2025-08-04 18:42:00	41	41.5000
+girl	2025-08-04 18:42:00	63	63.5000
+girl	2025-08-04 18:42:00	28	28.5000
+girl	2025-08-04 18:42:00	72	72.5000
+girl	2025-08-04 18:42:00	61	61.5000
+girl	2025-08-04 18:42:00	81	81.5000
+girl	2025-08-04 18:42:00	86	86.5000
+girl	2025-08-04 18:42:00	79	79.5000
+girl	2025-08-04 18:42:00	77	77.5000
+girl	2025-08-04 18:42:00	88	88.5000
+girl	2025-08-04 18:42:00	90	90.5000
+girl	2025-08-04 18:42:00	17	17.5000
+girl	2025-08-04 18:42:00	15	15.5000
+girl	2025-08-04 18:42:00	13	13.5000
+girl	2025-08-04 18:42:00	59	59.5000
+girl	2025-08-04 18:42:00	8	8.5000
+girl	2025-08-04 18:42:00	4	4.5000
+secret	2025-08-04 18:42:00	91	87.0000
+secret	2025-08-04 18:42:00	18	14.0000
+secret	2025-08-04 18:42:00	33	17.0000
+secret	2025-08-04 18:42:00	9	17.0000
+secret	2025-08-04 18:42:00	42	38.0000
+secret	2025-08-04 18:42:00	64	69.0000
+secret	2025-08-04 18:42:00	82	69.0000
+secret	2025-08-04 18:42:00	73	69.0000
+show status like '%Cached_plan%';
+Variable_name	Value
+Cached_plan_count	1
+Cached_plan_hits	12
+DROP PREPARE stmt1;
+DROP TABLE t_general_test_pk;
 set rds_plan_cache = @old_plan_cache;
diff --git a/mysql-test/suite/dstore_main/r/session_plan_cache.result b/mysql-test/suite/dstore_main/r/session_plan_cache.result
index c0d3c8421d0..b531f1e1c91 100644
--- a/mysql-test/suite/dstore_main/r/session_plan_cache.result
+++ b/mysql-test/suite/dstore_main/r/session_plan_cache.result
@@ -2650,4 +2650,494 @@ Warnings:
 Note	1003	/* select#1 */ select `test`.`t2`.`pk` AS `pk`,`test`.`t2`.`a` AS `a`,`test`.`t2`.`b` AS `b` from `test`.`t2` where (`test`.`t2`.`a` = 3)
 drop prepare s;
 drop table t2;
+#
+# BUGFIX while using windown function and order by together, after enable plan_cache, the query using cached plan will return a result set of incorrectly sorted
+#
+CREATE TABLE t_general_test_pk (
+id int NOT NULL auto_increment primary key,
+name varchar(255) DEFAULT NULL,
+client char(4) DEFAULT NULL,
+bin_id binary(16) NOT NULL,
+t_bit16 bit(16) DEFAULT NULL,
+create_time datetime DEFAULT CURRENT_TIMESTAMP COMMENT '????',
+col1 decimal(10,2) DEFAULT NULL,
+text blob,
+tex text,
+sex enum('boy','girl','secret') DEFAULT 'secret'
+) DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
+insert into t_general_test_pk values(1,null,null,123,0,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(2,null,null,123,0,'2025-08-04 18:42:00',0,null,null,'boy');
+insert into t_general_test_pk values(3,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(4,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'','','girl');
+insert into t_general_test_pk values(5,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(6,'NULL','NULL',123,0,'2025-08-04 18:42:00',0,'','','girl');
+insert into t_general_test_pk values(7,'NULL','NULL',123,2,'2025-08-04 18:42:00',0,'textsefwefwef','fwgfwergwegf','boy');
+insert into t_general_test_pk values(8,'NULL','NULL',123,2,'2025-08-04 18:42:00',0,'','','girl');
+insert into t_general_test_pk values(9,'NULL','NULL',123,2,'2025-08-04 18:42:00',0,'','','secret');
+insert into t_general_test_pk(name,client,bin_id,t_bit16,create_time,col1,text,tex,sex) select name,client,bin_id,t_bit16,create_time,col1,text,tex,sex from t_general_test_pk;
+insert into t_general_test_pk(name,client,bin_id,t_bit16,create_time,col1,text,tex,sex) select name,client,bin_id,t_bit16,create_time,col1,text,tex,sex from t_general_test_pk;
+insert into t_general_test_pk(name,client,bin_id,t_bit16,create_time,col1,text,tex,sex) select name,client,bin_id,t_bit16,create_time,col1,text,tex,sex from t_general_test_pk;
+create index id_key on t_general_test_pk(id desc);
+create index name_key on t_general_test_pk(name(20) desc, create_time);
+create index decimal_key on t_general_test_pk(col1);
+create index sex_key on t_general_test_pk(sex desc);
+set @sql_text1='select sex,create_time,id,id- LAG(id) over (order by create_time) as temp_diff_id from t_general_test_pk t order by sex,create_time;';
+PREPARE stmt1 FROM @sql_text1;
+EXECUTE stmt1;
+sex	create_time	id	temp_diff_id
+boy	2025-08-04 18:42:00	5	-1
+boy	2025-08-04 18:42:00	56	-28
+boy	2025-08-04 18:42:00	80	-1
+boy	2025-08-04 18:42:00	78	-1
+boy	2025-08-04 18:42:00	76	-1
+boy	2025-08-04 18:42:00	75	-1
+boy	2025-08-04 18:42:00	65	NULL
+boy	2025-08-04 18:42:00	16	-1
+boy	2025-08-04 18:42:00	14	-1
+boy	2025-08-04 18:42:00	12	-1
+boy	2025-08-04 18:42:00	11	-1
+boy	2025-08-04 18:42:00	25	14
+boy	2025-08-04 18:42:00	7	-1
+boy	2025-08-04 18:42:00	10	-65
+boy	2025-08-04 18:42:00	3	-1
+boy	2025-08-04 18:42:00	2	-1
+boy	2025-08-04 18:42:00	34	32
+boy	2025-08-04 18:42:00	40	-1
+boy	2025-08-04 18:42:00	38	-1
+boy	2025-08-04 18:42:00	36	-1
+boy	2025-08-04 18:42:00	35	-1
+boy	2025-08-04 18:42:00	1	-34
+boy	2025-08-04 18:42:00	31	-1
+boy	2025-08-04 18:42:00	29	-1
+boy	2025-08-04 18:42:00	27	-1
+boy	2025-08-04 18:42:00	26	-1
+boy	2025-08-04 18:42:00	67	-1
+boy	2025-08-04 18:42:00	62	-1
+boy	2025-08-04 18:42:00	84	-1
+boy	2025-08-04 18:42:00	60	-1
+boy	2025-08-04 18:42:00	74	8
+boy	2025-08-04 18:42:00	58	-1
+boy	2025-08-04 18:42:00	57	-1
+boy	2025-08-04 18:42:00	83	26
+boy	2025-08-04 18:42:00	66	-1
+boy	2025-08-04 18:42:00	69	-1
+boy	2025-08-04 18:42:00	89	-1
+boy	2025-08-04 18:42:00	87	-1
+boy	2025-08-04 18:42:00	71	-1
+boy	2025-08-04 18:42:00	85	-1
+girl	2025-08-04 18:42:00	68	-1
+girl	2025-08-04 18:42:00	6	-1
+girl	2025-08-04 18:42:00	37	-1
+girl	2025-08-04 18:42:00	32	-1
+girl	2025-08-04 18:42:00	70	-1
+girl	2025-08-04 18:42:00	39	-1
+girl	2025-08-04 18:42:00	30	-1
+girl	2025-08-04 18:42:00	41	-1
+girl	2025-08-04 18:42:00	63	-1
+girl	2025-08-04 18:42:00	28	-1
+girl	2025-08-04 18:42:00	72	-1
+girl	2025-08-04 18:42:00	61	-1
+girl	2025-08-04 18:42:00	81	-1
+girl	2025-08-04 18:42:00	86	-1
+girl	2025-08-04 18:42:00	79	-1
+girl	2025-08-04 18:42:00	77	-1
+girl	2025-08-04 18:42:00	88	-1
+girl	2025-08-04 18:42:00	90	-1
+girl	2025-08-04 18:42:00	17	-1
+girl	2025-08-04 18:42:00	15	-1
+girl	2025-08-04 18:42:00	13	-1
+girl	2025-08-04 18:42:00	59	-1
+girl	2025-08-04 18:42:00	8	-1
+girl	2025-08-04 18:42:00	4	-1
+secret	2025-08-04 18:42:00	91	8
+secret	2025-08-04 18:42:00	18	8
+secret	2025-08-04 18:42:00	33	32
+secret	2025-08-04 18:42:00	9	-16
+secret	2025-08-04 18:42:00	42	8
+secret	2025-08-04 18:42:00	64	-10
+secret	2025-08-04 18:42:00	82	26
+secret	2025-08-04 18:42:00	73	8
+EXECUTE stmt1;
+sex	create_time	id	temp_diff_id
+boy	2025-08-04 18:42:00	5	-1
+boy	2025-08-04 18:42:00	56	-28
+boy	2025-08-04 18:42:00	80	-1
+boy	2025-08-04 18:42:00	78	-1
+boy	2025-08-04 18:42:00	76	-1
+boy	2025-08-04 18:42:00	75	-1
+boy	2025-08-04 18:42:00	65	NULL
+boy	2025-08-04 18:42:00	16	-1
+boy	2025-08-04 18:42:00	14	-1
+boy	2025-08-04 18:42:00	12	-1
+boy	2025-08-04 18:42:00	11	-1
+boy	2025-08-04 18:42:00	25	14
+boy	2025-08-04 18:42:00	7	-1
+boy	2025-08-04 18:42:00	10	-65
+boy	2025-08-04 18:42:00	3	-1
+boy	2025-08-04 18:42:00	2	-1
+boy	2025-08-04 18:42:00	34	32
+boy	2025-08-04 18:42:00	40	-1
+boy	2025-08-04 18:42:00	38	-1
+boy	2025-08-04 18:42:00	36	-1
+boy	2025-08-04 18:42:00	35	-1
+boy	2025-08-04 18:42:00	1	-34
+boy	2025-08-04 18:42:00	31	-1
+boy	2025-08-04 18:42:00	29	-1
+boy	2025-08-04 18:42:00	27	-1
+boy	2025-08-04 18:42:00	26	-1
+boy	2025-08-04 18:42:00	67	-1
+boy	2025-08-04 18:42:00	62	-1
+boy	2025-08-04 18:42:00	84	-1
+boy	2025-08-04 18:42:00	60	-1
+boy	2025-08-04 18:42:00	74	8
+boy	2025-08-04 18:42:00	58	-1
+boy	2025-08-04 18:42:00	57	-1
+boy	2025-08-04 18:42:00	83	26
+boy	2025-08-04 18:42:00	66	-1
+boy	2025-08-04 18:42:00	69	-1
+boy	2025-08-04 18:42:00	89	-1
+boy	2025-08-04 18:42:00	87	-1
+boy	2025-08-04 18:42:00	71	-1
+boy	2025-08-04 18:42:00	85	-1
+girl	2025-08-04 18:42:00	68	-1
+girl	2025-08-04 18:42:00	6	-1
+girl	2025-08-04 18:42:00	37	-1
+girl	2025-08-04 18:42:00	32	-1
+girl	2025-08-04 18:42:00	70	-1
+girl	2025-08-04 18:42:00	39	-1
+girl	2025-08-04 18:42:00	30	-1
+girl	2025-08-04 18:42:00	41	-1
+girl	2025-08-04 18:42:00	63	-1
+girl	2025-08-04 18:42:00	28	-1
+girl	2025-08-04 18:42:00	72	-1
+girl	2025-08-04 18:42:00	61	-1
+girl	2025-08-04 18:42:00	81	-1
+girl	2025-08-04 18:42:00	86	-1
+girl	2025-08-04 18:42:00	79	-1
+girl	2025-08-04 18:42:00	77	-1
+girl	2025-08-04 18:42:00	88	-1
+girl	2025-08-04 18:42:00	90	-1
+girl	2025-08-04 18:42:00	17	-1
+girl	2025-08-04 18:42:00	15	-1
+girl	2025-08-04 18:42:00	13	-1
+girl	2025-08-04 18:42:00	59	-1
+girl	2025-08-04 18:42:00	8	-1
+girl	2025-08-04 18:42:00	4	-1
+secret	2025-08-04 18:42:00	91	8
+secret	2025-08-04 18:42:00	18	8
+secret	2025-08-04 18:42:00	33	32
+secret	2025-08-04 18:42:00	9	-16
+secret	2025-08-04 18:42:00	42	8
+secret	2025-08-04 18:42:00	64	-10
+secret	2025-08-04 18:42:00	82	26
+secret	2025-08-04 18:42:00	73	8
+EXECUTE stmt1;
+sex	create_time	id	temp_diff_id
+boy	2025-08-04 18:42:00	5	-1
+boy	2025-08-04 18:42:00	56	-28
+boy	2025-08-04 18:42:00	80	-1
+boy	2025-08-04 18:42:00	78	-1
+boy	2025-08-04 18:42:00	76	-1
+boy	2025-08-04 18:42:00	75	-1
+boy	2025-08-04 18:42:00	65	NULL
+boy	2025-08-04 18:42:00	16	-1
+boy	2025-08-04 18:42:00	14	-1
+boy	2025-08-04 18:42:00	12	-1
+boy	2025-08-04 18:42:00	11	-1
+boy	2025-08-04 18:42:00	25	14
+boy	2025-08-04 18:42:00	7	-1
+boy	2025-08-04 18:42:00	10	-65
+boy	2025-08-04 18:42:00	3	-1
+boy	2025-08-04 18:42:00	2	-1
+boy	2025-08-04 18:42:00	34	32
+boy	2025-08-04 18:42:00	40	-1
+boy	2025-08-04 18:42:00	38	-1
+boy	2025-08-04 18:42:00	36	-1
+boy	2025-08-04 18:42:00	35	-1
+boy	2025-08-04 18:42:00	1	-34
+boy	2025-08-04 18:42:00	31	-1
+boy	2025-08-04 18:42:00	29	-1
+boy	2025-08-04 18:42:00	27	-1
+boy	2025-08-04 18:42:00	26	-1
+boy	2025-08-04 18:42:00	67	-1
+boy	2025-08-04 18:42:00	62	-1
+boy	2025-08-04 18:42:00	84	-1
+boy	2025-08-04 18:42:00	60	-1
+boy	2025-08-04 18:42:00	74	8
+boy	2025-08-04 18:42:00	58	-1
+boy	2025-08-04 18:42:00	57	-1
+boy	2025-08-04 18:42:00	83	26
+boy	2025-08-04 18:42:00	66	-1
+boy	2025-08-04 18:42:00	69	-1
+boy	2025-08-04 18:42:00	89	-1
+boy	2025-08-04 18:42:00	87	-1
+boy	2025-08-04 18:42:00	71	-1
+boy	2025-08-04 18:42:00	85	-1
+girl	2025-08-04 18:42:00	68	-1
+girl	2025-08-04 18:42:00	6	-1
+girl	2025-08-04 18:42:00	37	-1
+girl	2025-08-04 18:42:00	32	-1
+girl	2025-08-04 18:42:00	70	-1
+girl	2025-08-04 18:42:00	39	-1
+girl	2025-08-04 18:42:00	30	-1
+girl	2025-08-04 18:42:00	41	-1
+girl	2025-08-04 18:42:00	63	-1
+girl	2025-08-04 18:42:00	28	-1
+girl	2025-08-04 18:42:00	72	-1
+girl	2025-08-04 18:42:00	61	-1
+girl	2025-08-04 18:42:00	81	-1
+girl	2025-08-04 18:42:00	86	-1
+girl	2025-08-04 18:42:00	79	-1
+girl	2025-08-04 18:42:00	77	-1
+girl	2025-08-04 18:42:00	88	-1
+girl	2025-08-04 18:42:00	90	-1
+girl	2025-08-04 18:42:00	17	-1
+girl	2025-08-04 18:42:00	15	-1
+girl	2025-08-04 18:42:00	13	-1
+girl	2025-08-04 18:42:00	59	-1
+girl	2025-08-04 18:42:00	8	-1
+girl	2025-08-04 18:42:00	4	-1
+secret	2025-08-04 18:42:00	91	8
+secret	2025-08-04 18:42:00	18	8
+secret	2025-08-04 18:42:00	33	32
+secret	2025-08-04 18:42:00	9	-16
+secret	2025-08-04 18:42:00	42	8
+secret	2025-08-04 18:42:00	64	-10
+secret	2025-08-04 18:42:00	82	26
+secret	2025-08-04 18:42:00	73	8
+show status like '%Cached_plan%';
+Variable_name	Value
+Cached_plan_count	1
+Cached_plan_hits	10
+DROP PREPARE stmt1;
+set @sql_text1='select sex,create_time,id,avg(id) over (order by create_time ROWS BETWEEN 1 PRECEDING AND CURRENT ROW) as moving_avg_id from t_general_test_pk t order by sex,create_time;';
+PREPARE stmt1 FROM @sql_text1;
+EXECUTE stmt1;
+sex	create_time	id	moving_avg_id
+boy	2025-08-04 18:42:00	5	5.5000
+boy	2025-08-04 18:42:00	56	70.0000
+boy	2025-08-04 18:42:00	80	80.5000
+boy	2025-08-04 18:42:00	78	78.5000
+boy	2025-08-04 18:42:00	76	76.5000
+boy	2025-08-04 18:42:00	75	75.5000
+boy	2025-08-04 18:42:00	65	65.0000
+boy	2025-08-04 18:42:00	16	16.5000
+boy	2025-08-04 18:42:00	14	14.5000
+boy	2025-08-04 18:42:00	12	12.5000
+boy	2025-08-04 18:42:00	11	11.5000
+boy	2025-08-04 18:42:00	25	18.0000
+boy	2025-08-04 18:42:00	7	7.5000
+boy	2025-08-04 18:42:00	10	42.5000
+boy	2025-08-04 18:42:00	3	3.5000
+boy	2025-08-04 18:42:00	2	2.5000
+boy	2025-08-04 18:42:00	34	18.0000
+boy	2025-08-04 18:42:00	40	40.5000
+boy	2025-08-04 18:42:00	38	38.5000
+boy	2025-08-04 18:42:00	36	36.5000
+boy	2025-08-04 18:42:00	35	35.5000
+boy	2025-08-04 18:42:00	1	18.0000
+boy	2025-08-04 18:42:00	31	31.5000
+boy	2025-08-04 18:42:00	29	29.5000
+boy	2025-08-04 18:42:00	27	27.5000
+boy	2025-08-04 18:42:00	26	26.5000
+boy	2025-08-04 18:42:00	67	67.5000
+boy	2025-08-04 18:42:00	62	62.5000
+boy	2025-08-04 18:42:00	84	84.5000
+boy	2025-08-04 18:42:00	60	60.5000
+boy	2025-08-04 18:42:00	74	70.0000
+boy	2025-08-04 18:42:00	58	58.5000
+boy	2025-08-04 18:42:00	57	57.5000
+boy	2025-08-04 18:42:00	83	70.0000
+boy	2025-08-04 18:42:00	66	66.5000
+boy	2025-08-04 18:42:00	69	69.5000
+boy	2025-08-04 18:42:00	89	89.5000
+boy	2025-08-04 18:42:00	87	87.5000
+boy	2025-08-04 18:42:00	71	71.5000
+boy	2025-08-04 18:42:00	85	85.5000
+girl	2025-08-04 18:42:00	68	68.5000
+girl	2025-08-04 18:42:00	6	6.5000
+girl	2025-08-04 18:42:00	37	37.5000
+girl	2025-08-04 18:42:00	32	32.5000
+girl	2025-08-04 18:42:00	70	70.5000
+girl	2025-08-04 18:42:00	39	39.5000
+girl	2025-08-04 18:42:00	30	30.5000
+girl	2025-08-04 18:42:00	41	41.5000
+girl	2025-08-04 18:42:00	63	63.5000
+girl	2025-08-04 18:42:00	28	28.5000
+girl	2025-08-04 18:42:00	72	72.5000
+girl	2025-08-04 18:42:00	61	61.5000
+girl	2025-08-04 18:42:00	81	81.5000
+girl	2025-08-04 18:42:00	86	86.5000
+girl	2025-08-04 18:42:00	79	79.5000
+girl	2025-08-04 18:42:00	77	77.5000
+girl	2025-08-04 18:42:00	88	88.5000
+girl	2025-08-04 18:42:00	90	90.5000
+girl	2025-08-04 18:42:00	17	17.5000
+girl	2025-08-04 18:42:00	15	15.5000
+girl	2025-08-04 18:42:00	13	13.5000
+girl	2025-08-04 18:42:00	59	59.5000
+girl	2025-08-04 18:42:00	8	8.5000
+girl	2025-08-04 18:42:00	4	4.5000
+secret	2025-08-04 18:42:00	91	87.0000
+secret	2025-08-04 18:42:00	18	14.0000
+secret	2025-08-04 18:42:00	33	17.0000
+secret	2025-08-04 18:42:00	9	17.0000
+secret	2025-08-04 18:42:00	42	38.0000
+secret	2025-08-04 18:42:00	64	69.0000
+secret	2025-08-04 18:42:00	82	69.0000
+secret	2025-08-04 18:42:00	73	69.0000
+EXECUTE stmt1;
+sex	create_time	id	moving_avg_id
+boy	2025-08-04 18:42:00	5	5.5000
+boy	2025-08-04 18:42:00	56	70.0000
+boy	2025-08-04 18:42:00	80	80.5000
+boy	2025-08-04 18:42:00	78	78.5000
+boy	2025-08-04 18:42:00	76	76.5000
+boy	2025-08-04 18:42:00	75	75.5000
+boy	2025-08-04 18:42:00	65	65.0000
+boy	2025-08-04 18:42:00	16	16.5000
+boy	2025-08-04 18:42:00	14	14.5000
+boy	2025-08-04 18:42:00	12	12.5000
+boy	2025-08-04 18:42:00	11	11.5000
+boy	2025-08-04 18:42:00	25	18.0000
+boy	2025-08-04 18:42:00	7	7.5000
+boy	2025-08-04 18:42:00	10	42.5000
+boy	2025-08-04 18:42:00	3	3.5000
+boy	2025-08-04 18:42:00	2	2.5000
+boy	2025-08-04 18:42:00	34	18.0000
+boy	2025-08-04 18:42:00	40	40.5000
+boy	2025-08-04 18:42:00	38	38.5000
+boy	2025-08-04 18:42:00	36	36.5000
+boy	2025-08-04 18:42:00	35	35.5000
+boy	2025-08-04 18:42:00	1	18.0000
+boy	2025-08-04 18:42:00	31	31.5000
+boy	2025-08-04 18:42:00	29	29.5000
+boy	2025-08-04 18:42:00	27	27.5000
+boy	2025-08-04 18:42:00	26	26.5000
+boy	2025-08-04 18:42:00	67	67.5000
+boy	2025-08-04 18:42:00	62	62.5000
+boy	2025-08-04 18:42:00	84	84.5000
+boy	2025-08-04 18:42:00	60	60.5000
+boy	2025-08-04 18:42:00	74	70.0000
+boy	2025-08-04 18:42:00	58	58.5000
+boy	2025-08-04 18:42:00	57	57.5000
+boy	2025-08-04 18:42:00	83	70.0000
+boy	2025-08-04 18:42:00	66	66.5000
+boy	2025-08-04 18:42:00	69	69.5000
+boy	2025-08-04 18:42:00	89	89.5000
+boy	2025-08-04 18:42:00	87	87.5000
+boy	2025-08-04 18:42:00	71	71.5000
+boy	2025-08-04 18:42:00	85	85.5000
+girl	2025-08-04 18:42:00	68	68.5000
+girl	2025-08-04 18:42:00	6	6.5000
+girl	2025-08-04 18:42:00	37	37.5000
+girl	2025-08-04 18:42:00	32	32.5000
+girl	2025-08-04 18:42:00	70	70.5000
+girl	2025-08-04 18:42:00	39	39.5000
+girl	2025-08-04 18:42:00	30	30.5000
+girl	2025-08-04 18:42:00	41	41.5000
+girl	2025-08-04 18:42:00	63	63.5000
+girl	2025-08-04 18:42:00	28	28.5000
+girl	2025-08-04 18:42:00	72	72.5000
+girl	2025-08-04 18:42:00	61	61.5000
+girl	2025-08-04 18:42:00	81	81.5000
+girl	2025-08-04 18:42:00	86	86.5000
+girl	2025-08-04 18:42:00	79	79.5000
+girl	2025-08-04 18:42:00	77	77.5000
+girl	2025-08-04 18:42:00	88	88.5000
+girl	2025-08-04 18:42:00	90	90.5000
+girl	2025-08-04 18:42:00	17	17.5000
+girl	2025-08-04 18:42:00	15	15.5000
+girl	2025-08-04 18:42:00	13	13.5000
+girl	2025-08-04 18:42:00	59	59.5000
+girl	2025-08-04 18:42:00	8	8.5000
+girl	2025-08-04 18:42:00	4	4.5000
+secret	2025-08-04 18:42:00	91	87.0000
+secret	2025-08-04 18:42:00	18	14.0000
+secret	2025-08-04 18:42:00	33	17.0000
+secret	2025-08-04 18:42:00	9	17.0000
+secret	2025-08-04 18:42:00	42	38.0000
+secret	2025-08-04 18:42:00	64	69.0000
+secret	2025-08-04 18:42:00	82	69.0000
+secret	2025-08-04 18:42:00	73	69.0000
+EXECUTE stmt1;
+sex	create_time	id	moving_avg_id
+boy	2025-08-04 18:42:00	5	5.5000
+boy	2025-08-04 18:42:00	56	70.0000
+boy	2025-08-04 18:42:00	80	80.5000
+boy	2025-08-04 18:42:00	78	78.5000
+boy	2025-08-04 18:42:00	76	76.5000
+boy	2025-08-04 18:42:00	75	75.5000
+boy	2025-08-04 18:42:00	65	65.0000
+boy	2025-08-04 18:42:00	16	16.5000
+boy	2025-08-04 18:42:00	14	14.5000
+boy	2025-08-04 18:42:00	12	12.5000
+boy	2025-08-04 18:42:00	11	11.5000
+boy	2025-08-04 18:42:00	25	18.0000
+boy	2025-08-04 18:42:00	7	7.5000
+boy	2025-08-04 18:42:00	10	42.5000
+boy	2025-08-04 18:42:00	3	3.5000
+boy	2025-08-04 18:42:00	2	2.5000
+boy	2025-08-04 18:42:00	34	18.0000
+boy	2025-08-04 18:42:00	40	40.5000
+boy	2025-08-04 18:42:00	38	38.5000
+boy	2025-08-04 18:42:00	36	36.5000
+boy	2025-08-04 18:42:00	35	35.5000
+boy	2025-08-04 18:42:00	1	18.0000
+boy	2025-08-04 18:42:00	31	31.5000
+boy	2025-08-04 18:42:00	29	29.5000
+boy	2025-08-04 18:42:00	27	27.5000
+boy	2025-08-04 18:42:00	26	26.5000
+boy	2025-08-04 18:42:00	67	67.5000
+boy	2025-08-04 18:42:00	62	62.5000
+boy	2025-08-04 18:42:00	84	84.5000
+boy	2025-08-04 18:42:00	60	60.5000
+boy	2025-08-04 18:42:00	74	70.0000
+boy	2025-08-04 18:42:00	58	58.5000
+boy	2025-08-04 18:42:00	57	57.5000
+boy	2025-08-04 18:42:00	83	70.0000
+boy	2025-08-04 18:42:00	66	66.5000
+boy	2025-08-04 18:42:00	69	69.5000
+boy	2025-08-04 18:42:00	89	89.5000
+boy	2025-08-04 18:42:00	87	87.5000
+boy	2025-08-04 18:42:00	71	71.5000
+boy	2025-08-04 18:42:00	85	85.5000
+girl	2025-08-04 18:42:00	68	68.5000
+girl	2025-08-04 18:42:00	6	6.5000
+girl	2025-08-04 18:42:00	37	37.5000
+girl	2025-08-04 18:42:00	32	32.5000
+girl	2025-08-04 18:42:00	70	70.5000
+girl	2025-08-04 18:42:00	39	39.5000
+girl	2025-08-04 18:42:00	30	30.5000
+girl	2025-08-04 18:42:00	41	41.5000
+girl	2025-08-04 18:42:00	63	63.5000
+girl	2025-08-04 18:42:00	28	28.5000
+girl	2025-08-04 18:42:00	72	72.5000
+girl	2025-08-04 18:42:00	61	61.5000
+girl	2025-08-04 18:42:00	81	81.5000
+girl	2025-08-04 18:42:00	86	86.5000
+girl	2025-08-04 18:42:00	79	79.5000
+girl	2025-08-04 18:42:00	77	77.5000
+girl	2025-08-04 18:42:00	88	88.5000
+girl	2025-08-04 18:42:00	90	90.5000
+girl	2025-08-04 18:42:00	17	17.5000
+girl	2025-08-04 18:42:00	15	15.5000
+girl	2025-08-04 18:42:00	13	13.5000
+girl	2025-08-04 18:42:00	59	59.5000
+girl	2025-08-04 18:42:00	8	8.5000
+girl	2025-08-04 18:42:00	4	4.5000
+secret	2025-08-04 18:42:00	91	87.0000
+secret	2025-08-04 18:42:00	18	14.0000
+secret	2025-08-04 18:42:00	33	17.0000
+secret	2025-08-04 18:42:00	9	17.0000
+secret	2025-08-04 18:42:00	42	38.0000
+secret	2025-08-04 18:42:00	64	69.0000
+secret	2025-08-04 18:42:00	82	69.0000
+secret	2025-08-04 18:42:00	73	69.0000
+show status like '%Cached_plan%';
+Variable_name	Value
+Cached_plan_count	1
+Cached_plan_hits	12
+DROP PREPARE stmt1;
+DROP TABLE t_general_test_pk;
 set rds_plan_cache = @old_plan_cache;
diff --git a/sql/sql_plan_cache.cc b/sql/sql_plan_cache.cc
index c4329716b75..48c741f12a3 100644
--- a/sql/sql_plan_cache.cc
+++ b/sql/sql_plan_cache.cc
@@ -1442,6 +1442,7 @@ bool JOIN::shallow_clone(JOIN *orig) {
   // PLAN_CACHE_PORT this is used in Query_expression::optimize() so let's try
   // to not let it stay at 0.
   best_rowcount = orig->best_rowcount;
+  m_windows_sort = orig->m_windows_sort;
 
   return false;
 }
-- 
2.23.0

